### ISSUE TITLE
Cannot execute beakerlib library self-tests

### ISSUE THREAD

**Issue by t184256** [teemtee/tmt#391]
Beakerlib libraries usually have `lib.sh` to `rlImport` from tests and `runtest.sh` containing self-tests for the library. `runtest.sh` thus `rlImport`s the appropriate lib.sh, and, worst of all it does it by name.

For example, https://github.com/beakerlib/openssl/tree/master/certgen contains `lib.sh` and `runtest.sh` with `rlImport openssl/certgen`. My attempts to execute it with `tmt` end up with `rlImport: Could not find library openssl/certgen`, as the test itself is not on `rlImport` lookup path.

I also advise against solving that with self-dependencies in metadata, because that'd require some way to prevent version discrepancies between `lib.sh` and `runtest.sh`. My guess is that we'll have to resort to some trick to enable beakerlib tests with to `rlImport` their `lib.sh`.

**Comment by psss** [comment]
@sopos, do you think that this could/should be handled on the `beakerlib` side? Is there a way to instruct `beakerlib` to look at the current working directory as well when looking for libraries?

**Comment by sopos** [comment]
Yes I think so. It mnight be worked around using `BEAKERLIB_LIBRARY_PATH="${BEAKERLIB_LIBRARY_PATH:+$BEAKERLIB_LIBRARY_PATH:}$PWD/../../"`

Currently the search in libraries root tries following locations
* $BEAKERLIB_LIBRARY_PATH/$COMPONENT/Library/$LIBRARY/lib.sh
* $BEAKERLIB_LIBRARY_PATH/*/$COMPONENT/Library/$LIBRARY/lib.sh
* $BEAKERLIB_LIBRARY_PATH/$COMPONENT/$LIBRARY/lib.sh
* $BEAKERLIB_LIBRARY_PATH/libs/$COMPONENT/$LIBRARY/lib.sh
During upwards traversal following are tried
* $DIRECTORY/$COMPONENT/Library/$LIBRARY/lib.sh
* $DIRECTORY/*/$COMPONENT/Library/$LIBRARY/lib.sh
* $DIRECTORY/libs/$COMPONENT/$LIBRARY/lib.sh

In the traversal the $DIRECTORY/$COMPONENT/$LIBRARY/lib.sh is missing. Adding it would solve this issue, I guess.

**Comment by psss** [comment]
Thanks for the detailed description, @sopos. Thinking about it a bit more: Would it make sense for `beakerlib` to always look at the current working directory? If it founds `lib.sh` and if it defines the requested library name it could directly import it without additional variables hacking needed. I think that library and its `runtest.sh` is are always closely related so such behaviour sounds natural to me.

**Comment by sopos** [comment]
There's a problem with it. There's no defined way how the library should specify its own name but directory structure. So beakerlib cannot tell what's the library in current WD without looking at the directories above.

**Comment by psss** [comment]
I see. So what about checking for `$cwd/../../$component/$library/lib.sh` and if it is found then use it?

**Comment by sopos** [comment]
That's basically the same as extending the upwards traversal with `$DIRECTORY/$COMPONENT/$LIBRARY/lib.sh`. I thing to check the same path patterns for _search in root_ and _upwards traveral_ makes sense well. I not think there's high chance of collisions.

**Comment by t184256** [comment]
I'm reopening that after a discussion with @sopos

This still doesn't work for me:

```
$ mkdir library-self-test && cd library-self-test
$ fmf init
$ beaker-wizard -s library  # and press Return many times to use defaults
$ cd Library/a-few-descriptive-words/
$ grep rlImport runtest.sh
        rlRun "rlImport library-self-test/a-few-descriptive-words"
$ tmt test import .
$ tmt run -a provision --how container  # fails
```

From the logs I can see that cwd is `/var/tmp/tmt/run-012/plans/default/discover/default/tests/Library/a-few-descriptive-words` and that rlImport fails with `[  ERROR   ] :: rlImport: Could not find library library-self-test/a-few-descriptive-words`. beakerlib's at `beakerlib-1.25-1.fc33`.

Note how the cwd doesn't contain the "component component" (`library-self-test`) at all, which means beakerlib can't use it when looking for `library-self-test/a-few-descriptive-words`.

**Comment by sopos** [comment]
@psss had an idea to use `rlImport .` for the case of self-import. It actually does not work in beakerlib at this point but it is really simple to add it.

However, what currently work is `rlImport ./a-few-descriptive-words` - you need to explicitly state the name of the library (the current directory name).

**Comment by psss** [comment]
The beakerlib support for `rlImport .` filed as beakerlib/beakerlib/issues/94.

**Comment by psss** [comment]
No update for more than 4 years, closing. Feel free to reopen if you still run into the problem.
